### PR TITLE
feat: support reading/writing Parquet datasets (#780)

### LINKED ISSUE #780: Support reading/writing Parquet datasets
## Summary

Add support for **reading (import) and writing (download/export) ORD datasets in Apache Parquet format**, alongside the existing protobuf serializations.

## Motivation

Parquet is a columnar, compressed format that's much more efficient than `.pb`/`.pbtxt`/`.json` for large datasets and integrates well with the data-science tooling ORD users rely on (pandas, Arrow, etc.). `ord-schema` has been growing first-class Parquet support for datasets, so exposing it in the app lets users round-trip datasets in the format they actually analyze in.

## Scope

- **Read / import:** accept a Parquet dataset file in the "Create dataset from file" / extend flows.
- **Write / export:** offer Parquet as a download format for datasets.

## Likely prerequisite: ord-schema version bump

This will probably require bumping **`ord-schema`** (currently `>=0.3.96,<0.4` in `pyproject.toml`) and **`ord-schema-protobufjs`** (currently `^0.3.96` in `ui/package.json`) to a version that ships the Parquet read/write helpers. The bump should be validated against the existing protobuf round-trip tests, since a minor/major ord-schema change can shift validation or proto field behavior.

## Technical pointers

- Backend serialization lives in [`ord_app/service_api/services/pb_utils.py`](ord_app/service_api/services/pb_utils.py): the `EXTENSION_*` kind maps (`.json`→json, `.pb`→binpb, `.pbtxt`→txtpb), `validate_pb_kind_by_file_ext` (upload), and `write_message` (download). Parquet would be a new "kind" handled here (note: Parquet is dataset-level, not a single-message serialization, so `write_message`/`validate_uploaded_pb_file` may need a dataset-specific path rather than the per-message kind switch).
- Frontend download options: `fileDownloadOptions` in `common/constants.ts` (drives the "Download as" menu); upload accepted extensions in the file-control/create-dataset flows.

## Acceptance criteria

1. A user can upload/import a `.parquet` dataset file and have its reactions created.
2. A user can download a dataset as `.parquet`, and that file round-trips back through import.
3. `ord-schema` / `ord-schema-protobufjs` are bumped to a Parquet-capable version with the existing test suites green.
4. Clear error handling for malformed/oversized Parquet uploads (consistent with the existing upload-limit and error-code behavior).

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### PR DESCRIPTION
**Pull request by skearnes**
Closes #780.

Adds **Apache Parquet** as a dataset-level import/export format alongside the existing `.binpb`/`.txtpb`/`.json` serializations, using `ord_schema.parquet` (shipped by the 0.6.x bump in #798).

## Backend
- **`pb_utils`**: new `MAP_FILE_EXT_TO_DATASET_KIND` (= the per-message map + `.parquet`) and dataset-level helpers `load_dataset_message` / `write_dataset_message`. Parquet is path-based in ord-schema, so the helpers stage bytes through a temp directory; non-Parquet kinds delegate to the existing per-message `load_message`/`write_message`. `validate_uploaded_pb_file` / `validate_pb_kind_by_file_ext` now take the extension map as a parameter.
- **Upload / extend** (`/datasets/upload`, `/datasets/{id}/extend`): accept `.parquet`. A malformed Parquet file (pyarrow error) or a valid Parquet that isn't an ORD dataset surfaces as a **400**, matching the other formats.
- **Download** (`/datasets/{id}/download`): adds `parquet`. ord-schema's writer requires a non-empty **description** and **≥1 reaction**, so export pre-checks both and returns a clear **422** (`"Parquet export requires a dataset description…"`) instead of a bare error. Single-reaction downloads keep `DownloadFileFormats`; datasets use the new `DatasetDownloadFileFormats` (adds `parquet`) so the dataset-only format isn't offered for reactions.
- `pyarrow` is now an explicit dependency (it was already transitive via ord-schema; app code imports it directly for the exception type).

## Frontend
- `datasetFileDownloadOptions` (= `fileDownloadOptions` + `.parquet`), wired into the **dataset header** and **dataset table** download menus only — reaction menus are unchanged.
- "Create dataset from file" accepts `.parquet`.

## Tests
- **Unit** (`test_pb_utils.py`): Parquet round-trip preserves name/description/reactions byte-for-byte; non-Parquet kinds delegate; malformed Parquet raises; empty dataset rejected; `.parquet` valid only under the dataset map.
- **API**: upload `.parquet`, malformed `.parquet` → 400, download `.parquet` round-trips back through the loader, and description-less export → 422.

## Acceptance criteria
1. ✅ Upload/import a `.parquet` dataset → reactions created.
2. ✅ Download a dataset as `.parquet` that round-trips back through import (covered by a test).
3. ✅ ord-schema/-protobufjs already on a Parquet-capable version (#798); full suites green.
4. ✅ Clear error handling for malformed (400) and non-exportable (422) cases.

## Decision note
Per maintainer call: a description-less dataset is **blocked with a clear 422** on `.parquet` export rather than having a synthetic description injected — keeps invented metadata out of exported scientific datasets. Other formats are unaffected.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

<h3>Greptile Summary</h3>

This PR adds Apache Parquet as a dataset-level import/export format, complementing the existing `.binpb`/`.txtpb`/`.json` serializations. It introduces path-based staging helpers (`load_dataset_message`/`write_dataset_message`) backed by a safe `tempfile.mkstemp` context manager, extends upload/extend/download routes to accept Parquet, surfaces clear 422s for missing description or zero reactions, and adds `DatasetDownloadFileFormats` to keep Parquet out of single-reaction menus.

- **Backend**: New `pb_utils` helpers stage Parquet through a temp directory; the download path pre-checks description and reaction count before handing off to ord-schema's writer; `ArrowException` and `ValueError` are caught alongside proto decode errors to return consistent 400s on bad uploads.
- **Frontend**: `datasetFileDownloadOptions` extends the existing constant with `.parquet` and is wired into the dataset header and table download menus only; the file-upload input now accepts `.parquet`.
- **Tests**: Round-trip, malformed-upload, and parametrised description-422 tests are provided at both the unit and API levels; the one minor gap is an API-level test for the zero-reactions 422 branch.

<h3>Confidence Score: 5/5</h3>

Safe to merge — the change is well-contained, error paths are explicitly handled, and temp-file cleanup is correct.

All new code paths (upload, extend, download) are exercised by tests, error responses match the rest of the API conventions, and the description/reaction pre-checks prevent malformed Parquet exports from producing bare server errors. The only gap is a missing API-level test for the zero-reactions 422 branch; everything else is covered end-to-end.

`datasets_test.py` — the zero-reactions Parquet download 422 path is untested at the API level.

<details><summary><h3>Important Files Changed</h3></summary>

| Filename | Overview |
|----------|----------|
| ord_app/service_api/services/pb_utils.py | Adds `load_dataset_message`/`write_dataset_message` + `_staged_parquet_path` context manager for temp-file staging; extends the two `validate_*` helpers to accept a pluggable `ext_to_kind` map; logic and cleanup handling look solid. |
| ord_app/service_api/domain/datasets.py | Switches `extend`/`upload` to `load_dataset_message`, adds `ArrowException`/`ValueError` to the decode-error handler, and adds pre-flight 422 guards (no reactions, missing/whitespace description) before the Parquet write path. |
| ord_app/service_api/resources/v1/datasets_test.py | Adds upload (valid + malformed), download round-trip, and parametrised description-422 API tests; no API-level test for the zero-reactions 422 branch. |
| ord_app/service_api/services/pb_utils_test.py | New unit test file covers Parquet round-trip, non-Parquet delegation, malformed input, and no-reactions/no-description rejection — thorough for the helpers layer. |
| ord_app/service_api/schemas/datasets.py | Adds `DatasetDownloadFileFormats` as a superset of `DownloadFileFormats`; `DownloadFileFormats` is preserved and still used by the reactions API. |
| ord_app/service_api/resources/v1/datasets.py | Threads `MAP_FILE_EXT_TO_DATASET_KIND` through `validate_uploaded_pb_file` for upload and extend routes; switches download to `DatasetDownloadFileFormats`. |
| ui/src/common/constants.ts | Adds `datasetFileDownloadOptions` as `fileDownloadOptions` + `.parquet`; the separation between dataset-only and reaction formats is clean. |
| pyproject.toml | Promotes `pyarrow` from a transitive to an explicit dependency with a `>=24,<25` pin; matches what ord-schema already requires. |

</details>

</details>

<details><summary><h3>Comments Outside Diff (1)</h3></summary>

1. `ord_app/tests/tests_datasets/test_create_dataset.py`, line 194-221 ([link](https://github.com/open-reaction-database/ord-app/blob/f2e0109bfa670b98aae4a56bfe210137ad343113/ord_app/tests/tests_datasets/test_create_dataset.py#L194-L221)) 

   <a href="#">P2</a> **No test for extending a dataset with Parquet**

   The extend path (`/datasets/{id}/extend`) is wired to accept `.parquet` files via `MAP_FILE_EXT_TO_DATASET_KIND`, but there is no test that exercises this route with a Parquet payload. The upload and download paths are well-covered; the extend path's Parquet branch (calling `load_dataset_message` with `kind="parquet"`) goes untested. A simple test analogous to `test_dataset_extend` but uploading `_parquet_dataset_bytes(...)` would provide confidence that the reactions are merged correctly and that malformed Parquet is rejected with a 400.

   Note: If this suggestion doesn't match your team's coding style, reply to this and let me know. I'll remember it for next time!
</details>

<sub>Reviews (3): Last reviewed commit: ["Merge remote-tracking branch &#39;origin/mai..."](https://github.com/open-reaction-database/ord-app/commit/604c67fcd2032a499b27f45833563aff81f59b57) | [Re-trigger Greptile](https://app.greptile.com/api/retrigger?id=38948499)</sub>

### CODE REVIEW

File: ord_app/service_api/domain/datasets.py [hunk 1]
```diff
@@ -254,7 +273,22 @@ async def download(
...
 
-        data = await run_in_threadpool(write_message, dataset_pb, kind=file_format)
+        if file_format == "parquet":
+            # ord-schema's Parquet writer requires a non-empty description and at least one
+            # reaction; surface a clear 422 rather than letting it raise a bare ValueError.
+            if not dataset_pb.reactions:
+                raise UnprocessableEntityError(
+                    "Parquet export requires at least one reaction in the dataset."
+                )
+            if not dataset_pb.description:
```
[the remark this answers is not in the document]
Comment (skearnes) [reply ord_app/service_api/domain/datasets.py]: Good catch on the intent. Tightened the guard to `if not dataset_pb.description.strip()` and extended the test to cover `None` / `""` / `"   "`.

One clarification: ord-schema's `DatasetWriter` uses a plain truthy check (`if not description`), so a whitespace-only `"   "` would actually have been *accepted* by `save_dataset` (no `ValueError`, no 500) and exported as a meaningless description — the stricter guard now rejects it with the intended 422.\n\n_Analysis performed with [Claude Code](https://claude.com/claude-code) (Claude Opus 4.8)._
